stdlib/public/runtime/Paths.cpp

```diff
@@ -0,0 +1,369 @@
+//
+// Functions that obtain paths that might be useful within the runtime.
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/Runtime/Config.h"
+#include "swift/Runtime/EnvironmentVariables.h"
+#include "swift/Runtime/Debug.h"
+#include "swift/Runtime/Paths.h"
+#include "swift/Runtime/Win32.h"
+#include "swift/Threading/Once.h"
+
+#include <filesystem>
+
+#if !defined(_WIN32) || defined(__CYGWIN__)
+#include <sys/stat.h>
+
+#include <dlfcn.h>
+#else
+#define WIN32_LEAN_AND_MEAN
+#define NOMINMAX
+#include <windows.h>
+#include <psapi.h>
+#endif
+
+#include <cerrno>
+#include <cstdlib>
+#include <cstring>
+
+namespace {
+
+swift::once_t runtimePathToken;
+const char *runtimePath;
+
+swift::once_t rootPathToken;
+const char *rootPath;
+
+void _swift_initRuntimePath(void *);
+void _swift_initRootPath(void *);
+const char *_swift_getDefaultRootPath();
+const char *_swift_getAuxExePathIn(const char *path, const char *name);
+
+bool _swift_isPathSep(char ch) {
+#ifdef _WIN32
+  return ch == '/' || ch == '\\';
+#else
+  return ch == '/';
+#endif
+}
+
+bool _swift_exists(const char *path);
+
+#if !defined(_WIN32) || defined(__CYGWIN__)
+#define PATHSEP_STR "/"
+#define PATHSEP_CHR '/'
+#else
+#define PATHSEP_STR "\\"
+#define PATHSEP_CHR '\\'
+#endif
+
+}
+
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getRuntimePath()
+{
+  swift::once(runtimePathToken, _swift_initRuntimePath, nullptr);
+  return runtimePath;
+}
+
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getRootPath()
+{
+  swift::once(rootPathToken, _swift_initRootPath, nullptr);
+  return rootPath;
+}
+
+namespace {
+
+const char *
+_swift_getDefaultRootPath()
+{
+  const char *runtimePath = swift_getRuntimePath();
+  size_t runtimePathLen = std::strlen(runtimePath);
+
+  // Scan backwards until we find a path separator
+  const char *ptr = runtimePath + runtimePathLen;
+  while (ptr > runtimePath && !_swift_isPathSep(*--ptr));
+
+  // Remove lib/swift/ if present
+  if (ptr - runtimePath >= 10
+      && _swift_isPathSep(ptr[-10])
+      && std::strncmp(ptr - 9, "lib", 3) == 0
+      && _swift_isPathSep(ptr[-6])
+      && std::strncmp(ptr - 5, "swift", 5) == 0) {
+    ptr -= 10;
+  }
+
+  // If the result is empty, return "./" or ".\\"
+  if (ptr == runtimePath) {
+    return "." PATHSEP_STR;
+  }
+
+  // Duplicate the string up to and including ptr
+  size_t len = ptr - runtimePath + 1;
+  char *thePath = (char *)malloc(len + 1);
+  std::memcpy(thePath, runtimePath, len);
+  thePath[len] = 0;
+
+  return thePath;
+}
+
+void
+_swift_initRootPath(void *)
+{
+  // SWIFT_ROOT overrides the path returned by this function
+  const char *swiftRoot = swift::runtime::environment::SWIFT_ROOT();
+  if (swiftRoot && *swiftRoot) {
+    size_t len = std::strlen(swiftRoot);
+
+    // Ensure that there's a trailing slash
+    if (_swift_isPathSep(swiftRoot[len - 1])) {
+      rootPath = swiftRoot;
+    } else {
+      char *thePath = (char *)malloc(len + 2);
```

Comment:
Yes. Using `new` in the runtime causes problems, because people can override `operator new` and since some of our code gets run really early, their `operator new` implementation might not be completely set-up at this point. We've had crashes caused by exactly that problem.

I don't think there's much benefit to using `strlcpy()` here, and I'm not sure how you'd use `strncat()` with `strdup()` (since the latter is going to allocate space just for the original string). `sprintf()` or `snprintf()` are an option too, but we already know the length of the string at this point, because we literally just measured it.